### ISSUE TITLE
Keras for many number of classes, e.g., language modelling with vocabulary size 50,000.

### ISSUE THREAD

**Issue by Palang2014** [fchollet/keras#1723]
Hi all,

First of all thanks for this wonderful framework and great work you guys have done.

I have been trying to use Keras to build a simple LSTM language model with a vocabulary of size 50,000. Considering one-hot representation of words, this translates into 50,000 classes. 
To avoid padding, I have been dividing my training data (~80 million sentences) into chunks where each chunk includes sentences of the same length. Model compiles well and training loss is decreasing but I have two main problems:

1) Using " `np_utils.to_categorical ()` " function results in huge memory (and computation?) waste. I tried "`K.mean(K.categorical_crossentropy(y_pred, K.cast(y_true.flatten(), dtype='int32')), axis=-1)`" to avoid it but got a rank mismatch error from Theano. Is there any work around for this? 

2) I guess when the number of classes is huge the running time becomes very high. Is there any solution for this?
E.g., in my language model experiment, with just one LSTM layer and a Softmax layer on the top, 512 cells for LSTM, maximum sentence length of 15, vocabulary size 50,000 and mini-batch size = 1024 sentences, it takes 4 sec / mini-batch on GPU, which is too long. 
It also consumes up to 7 GB of GPU memory which I believe is because of  " `np_utils.to_categorical ()` " function.

Thanks!
Hamid

**Comment by neggert** [comment]
I think the traditional approach to this is to use hierarchical softmax. Don't believe this is implemented in Keras (yet?).

**Comment by Palang2014** [comment]
@neggert : I thought that it is the hierarchical softmax by default. I just found out there is Advanced Activation Layers in Keras but it does not include hierarchical softmax. Updated my Keras installation and checked the layers, was not there as well. Guess it is not implemented yet ...

**Comment by Palang2014** [comment]
@xiaodonghe
@zer0n

**Comment by KeironO** [comment]
Someone has already implemented hierarchial softmax Theano, so it should just work with Keras.

https://github.com/RobGrimm/hierarchical_softmax/blob/master/HierarchicalSoftmax.py

**Comment by bkj** [comment]
@KeironO Could you elaborate on how to use this with Keras?  Do you just drop it in with a `Lambda` layer?

**Comment by KeironO** [comment]
Yes.

**Comment by farizrahman4u** [comment]
There are weights inside HSM, Lambda layers are for stateless functions.

**Comment by farizrahman4u** [comment]
I will be rolling out a very tightly optimised softmax layer this weekend. So hang on.

**Comment by erickrf** [comment]
@farizrahman4u any news on the hierarchical softmax?

**Comment by farizrahman4u** [comment]
github.com/farizrahman4u/huffmax
